ocaml/testsuite/tests/typing-layouts-arrays/basics.ml

```diff
@@ -0,0 +1,202 @@
+
+let v1 = [| #1. |]
+[%%expect{|
+val v1 : float# array = [|<abstr>|]
+|}];;
+
+
+let v2 = [| #1l |]
+[%%expect{|
+val v2 : int32# array = [|<abstr>|]
+|}];;
+
+
+let v3 = [| #1L |]
+[%%expect{|
+val v3 : int64# array = [|<abstr>|]
+|}];;
+
+
+let v4 = [| #1n |]
+[%%expect{|
+val v4 : nativeint# array = [|<abstr>|]
+|}];;
+
+(****************************************)
+(* Test 3: Array operations do not work *)
+
+let f (x : float# array) = x.(0)
+[%%expect{|
+Line 1, characters 27-28:
+1 | let f (x : float# array) = x.(0)
+                               ^
+Error: This expression has type float# array
+       but an expression was expected of type 'a array
+       The layout of float# is float64, because
+         it is the primitive float64 type float#.
+       But the layout of float# must be a sublayout of value, because
+         of layout requirements from an imported definition.
+|}];;
+
+let f (x : float# array) = Array.length x
+[%%expect{|
+Line 1, characters 40-41:
+1 | let f (x : float# array) = Array.length x
+                                            ^
+Error: This expression has type float# array
+       but an expression was expected of type 'a array
+       The layout of float# is float64, because
+         it is the primitive float64 type float#.
+       But the layout of float# must be a sublayout of value, because
+         of layout requirements from an imported definition.
+|}];;
+
+(*****************************************************************)
+(* Test 4: Calling wrong primitives on unboxed array kinds fails *)
+
+external get : float# array -> int -> float = "%floatarray_safe_get"
+let d (x : float# array) = get x 0
+
+[%%expect{|
+external get : float# array -> int -> float = "%floatarray_safe_get"
+Line 2, characters 27-34:
+2 | let d (x : float# array) = get x 0
+                               ^^^^^^^
+Error: Array kind unboxed_float can only be operated on using its own primitives
+       and those primitives can only work on unboxed_float
+|}];;
+
+
+(* Doesn't prevent the use of [Obj.magic] *)
+external get : floatarray -> int -> float = "%floatarray_safe_get"
+let d (x : float# array) = get (Obj.magic x : floatarray) 0
+
+[%%expect{|
+external get : floatarray -> int -> float = "%floatarray_safe_get"
+val d : float# array -> float = <fun>
+|}];;
+
+external get : ('a : any). 'a array -> int -> float = "%floatarray_safe_get"
+let d (x : 'a array) = get x 0
+
+[%%expect{|
+external get : ('a : any). 'a array -> int -> float = "%floatarray_safe_get"
+Line 2, characters 23-30:
+2 | let d (x : 'a array) = get x 0
+                           ^^^^^^^
+Error: A representable layout is required here.
+       The layout of 'a is any, because
+         of the definition of d at line 2, characters 6-30.
+       But the layout of 'a must be representable, because
+         it's the type of an array element.
+|}];;
+
+external get : int32# array -> int -> float = "%floatarray_safe_get"
+let d (x : int32# array) = get x 0
```

Comment:
It's hard to produce an error on every invalid external declaration without doing full type checking. The comment in `prim_has_valid_reprs` points out that limitation:

https://github.com/ocaml-flambda/flambda-backend/blob/96be100bbd003e834fc81de86ac69444ac8e2f6e/ocaml/typing/primitive.ml#L473-L486

The glb check at usage sites can catch some more mistakes but I believe it's also not comprehensive.

```
external get : int32# array -> int -> float = "%floatarray_safe_get"
```

produces an error, but not:

```
external get : float# array -> int -> int32# = "%array_safe_get" 
```

The glb check for this is between Pgenarray and Punboxedfloatarray. Nothing checks the `int32#`.

Reply:
I think maybe we can live with this, thanks to middle-end checks.  But without them I would be pretty worried.  It seems easy to make copy-paste errors of this class, and in particular:
```ocaml
external get : float# array -> int -> float = "%array_safe_get"
```
seems very easy to write on accident.

I'm wondering if it's worth trying to do some more type checking specifically for array primitives.  A thing I think we could do, somewhere like `prim_has_valid_reprs`, is when we see `array_safe_get` we unify the first argument with `('a : any) array` and then either get the layout on `'a` and check the return layout corresponds, or just unify the return type with `'a`.

I think this would rule out some mistakes.  I'm a little worried it would also rule out some existing programs in our code base - places where somebody is doing something funky with obj.magic, or is using a type alias for arrays that we can't expand for whatever reason.  Maybe worth trying though?

Reply:
Resolved by in-person discussion. Agreed that this is a pretty unlikely declaration for users to write. Filed internal ticket to be addressed later